docs/proposals/20191017-kubeadm-based-control-plane.md

```diff
@@ -351,41 +426,60 @@ spec:
+  - Changes to the infrastructureRef
+  - The `upgradeAfter` field, which can be set to a specific time in the future
+    - Set to `nil` or the zero value of `time.Time` if no upgrades are desired
+    - An upgrade will run after that timestamp is passed
+    - Good for scheduling upgrades/SLOs
+    - Set `upgradeAfter` to now (in RFC3339 form) if an upgrade is required immediately
+
+- The controller should tolerate the manual or automatic removal of a replica during the upgrade process. A replica that fails during the upgrade may block the completion of the upgrade. Removal or other remedial action may be necessary to allow the upgrade to complete.
+
+- In order to determine if a Machine to be rolled out, KCP implements the following:
+    - The infrastructureRef link used by each machine at creation time is stored in annotations at machine level.
+    - The kubeadmConfigSpec used by each machine at creation time is stored in annotations at machine level.
+        - If the annotation is not present (machine is either old or adopted), we won't roll out on any possible changes made in KCP's ClusterConfiguration given that we don't have enough information to make a decision. Users should use KCP.Spec.UpgradeAfter field to force a rollout in this case.
+
+##### Rolling update strategy
+
+Currently KubeadmControlPlane supports only one rollout strategy type the `RollingUpdateStrategyType`. Rolling upgrade strategy's behavior can be modified by using `MaxUnavailable` and `MaxSurge` fields. Both field values can be an absolute number 0 or 1 with following rules:
+
+- If `MaxUnavailable` is set to 0 `MaxSurge` needs to be 1 (default values)
+- If `MaxUnavailable` is set to 1 `MaxSurge` needs to be 0
+
+When `MaxUnavailable` is set to 0 and `MaxSurge` is set to 1 the rollout algorithm is as follows:  
+
   - Find Machines that have an outdated spec
   - If there is a machine requiring rollout
     - Scale up control plane creating a machine with the new spec
     - Scale down control plane by removing one of the machine that needs rollout (the oldest out-of date machine in the failure domain that has the most control-plane machines on it)
 
-- In order to determine if a Machine to be rolled out, KCP implements the following:
-    - The infrastructureRef link used by each machine at creation time is stored in annotations at machine level.
-    - The kubeadmConfigSpec used by each machine at creation time is stored in annotations at machine level.
-        - If the annotation is not present (machine is either old or adopted), we won't roll out on any possible changes made in KCP's ClusterConfiguration given that we don't have enough information to make a decision.
-           Users should use KCP.Spec.UpgradeAfter field to force a rollout in this case.
+Setting `MaxUnavailable` to 1 and `MaxSurge` to 0 is recommended for resource constrained environment like bare-metal, OpenStack or vSphere resource pools, etc. The algorithm is the following:
```

Comment:
I don't want to block progress on this, but as a followup can we refine this to avoid saying that it is recommended to use these settings, but rather that one *could* use these settings if they are operating in a resource constrained environment and do not have sufficient capacity to use a surge based rollout?

Reply:
+1